src/spec.rs

```diff
@@ -1,43 +1,206 @@
+#![allow(clippy::wrong_self_convention)]
+#![macro_use]
+
 use crate::result::{ZipError, ZipResult};
-use crate::unstable::{LittleEndianReadExt, LittleEndianWriteExt};
+use memchr::memmem::FinderRev;
 use std::borrow::Cow;
 use std::io;
 use std::io::prelude::*;
+use std::mem;
 use std::path::{Component, Path, MAIN_SEPARATOR};
 
-pub const LOCAL_FILE_HEADER_SIGNATURE: u32 = 0x04034b50;
-pub const CENTRAL_DIRECTORY_HEADER_SIGNATURE: u32 = 0x02014b50;
-pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06054b50;
-pub const ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06064b50;
-pub(crate) const ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE: u32 = 0x07064b50;
+pub type Magic = u32;
+
+pub const LOCAL_FILE_HEADER_SIGNATURE: Magic = 0x04034b50;
+pub const CENTRAL_DIRECTORY_HEADER_SIGNATURE: Magic = 0x02014b50;
+pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: Magic = 0x06054b50;
+pub const ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE: Magic = 0x06064b50;
+pub(crate) const ZIP64_CENTRAL_DIRECTORY_END_LOCATOR_SIGNATURE: Magic = 0x07064b50;
 
 pub const ZIP64_BYTES_THR: u64 = u32::MAX as u64;
 pub const ZIP64_ENTRY_THR: usize = u16::MAX as usize;
 
+pub trait Block: Sized + Copy {
+    /* TODO: use smallvec? */
+    fn interpret(bytes: Box<[u8]>) -> ZipResult<Self>;
+
+    fn deserialize(block: &[u8]) -> Self {
+        assert_eq!(block.len(), mem::size_of::<Self>());
+        let block_ptr: *const Self = block.as_ptr().cast();
+        unsafe { block_ptr.read() }
+    }
+
+    fn parse<T: Read>(reader: &mut T) -> ZipResult<Self> {
+        let mut block = vec![0u8; mem::size_of::<Self>()];
+        reader.read_exact(&mut block)?;
+        Self::interpret(block.into_boxed_slice())
+    }
+
+    fn encode(self) -> Box<[u8]>;
+
+    fn serialize(self) -> Box<[u8]> {
+        let mut out_block = vec![0u8; mem::size_of::<Self>()];
```

Comment:
I get this error if I try that:
```rust
error: constant expression depends on a generic parameter
  --> src/spec.rs:41:35
   |
41 |         let mut out_block = [0u8; mem::size_of::<Self>()];
   |                                   ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this may fail depending on what value the parameter takes
```

However, I have read on twitter that `realloc()` on most systems (which is what we get from `.into_boxed_slice()`) is essentially a no-op on most systems if the size is the same or smaller. So I think we don't lose much performance even if we have to dynamically allocate up front here.

Reply:
That's extremely clever, but unfortunately that requires us to `impl` *all* of the methods of `Block` within that block. We are able to avoid putting `Self` within the body of a constant value expression by using `MaybeUninit`, but that leads to a final issue:
```rust
error: generic parameters may not be used in const operations
  --> src/spec.rs:84:51
   |
84 |     fn serialize(self) -> Box<[u8; mem::size_of::<Self>()]> {
   |                                                   ^^^^ cannot perform const operation using `Self`
   |
   = note: type parameters may not be used in const expressions
   = help: add `#![feature(generic_const_exprs)]` to allow generic const expressions

error: generic parameters may not be used in const operations
  --> src/spec.rs:86:60
   |
86 |         let mut out_block: MaybeUninit<[u8; mem::size_of::<Self>()]> = MaybeUninit::zeroed();
   |                                                            ^^^^ cannot perform const operation using `Self`
   |
   = note: type parameters may not be used in const expressions
   = help: add `#![feature(generic_const_exprs)]` to allow generic const expressions
```

I'm going to keep this as is now, but add a note that it can be made to generate a boxed array directly with an unstable feature. I've also refactored to add `to_le()` and `from_le()` as methods of `Block` itself instead of the weird implicit mixin we had before that relied on the macro.

Reply:
It does not seem possible to `.try_into()` when using a trait like this, as without `generic_const_exprs`, `mem::size_of::<Self>()` cannot be used in `const` contexts. However, we are not actually performing any size checking except once in `deserialize()`, because we are casting the heap-allocated value and writing directly to a pointer: https://github.com/zip-rs/zip2/blob/70215c855cd667e1e8766f6aaa1ac4d339a2ba94/src/spec.rs#L116-L120 https://github.com/zip-rs/zip2/blob/70215c855cd667e1e8766f6aaa1ac4d339a2ba94/src/spec.rs#L138-L147

Because of this, I do not think we are adding any bounds checks except the explicit `assert_eq!()` in `deserialize()`. I'm sure that imparts some overhead, but it's not clear that it's too much overhead.

One way we could test this is to enable `#![feature(generic_const_exprs)]` and modify each method to use const array sizes, to see if that meaningfully improves our benchmarks. I think adopting `iai` benchmarks (which uses cachegrind to count instructions) would be a great way to validate this theory in a followup change. That might also allow us to avoid boxing entirely, so I think it's a good idea. While stable traits can't do `const` methods, we could potentially apply a `macro_rules!` macro to do the same.